### ISSUE TITLE
PartialEq on Vec fails with a Why3 type error

### ISSUE THREAD

**Issue by gmorenz** [xldenis/creusot#600]
I haven't really thought about why this is happening yet, I'll look at this in more detail tomorrow but thought I should document it now.

This code

```rust
use creusot_contracts::*;

#[ensures(result == true)]
fn veq() -> bool {
    let v1: Vec<i32> = Vec::new();
    let v2: Vec<i32> = Vec::new();
    v1 == v2
}

fn main() {}
```

Fails with this error

```
File "./target/debug/testbin-bin.mlcfg", line 165, characters 44-62:
This term has type seq.Seq.seq Alloc_Vec_PartialEq_Impl0_Eq_Interface.u,
but is expected to have type seq.Seq.seq
Alloc_Vec_PartialEq_Impl0_Eq_Interface.t
```

And this seems to be the relevant whyml snippet (line 165 is the `ensures` line at the end of the module).

```
module Alloc_Vec_PartialEq_Impl0_Eq_Interface
  type t
  type u
  type a1
  type a2
  use prelude.Borrow
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy_Type as ModelTy1 with type t = u, type a = a2
  clone CreusotContracts_Std1_Vec_Impl0_ModelTy_Type as ModelTy0 with type t = t, type a = a1
  use Alloc_Vec_Vec_Type as Alloc_Vec_Vec_Type
  clone CreusotContracts_Logic_Model_Impl0_Model_Stub as Model1 with type t = Alloc_Vec_Vec_Type.t_vec u a2,
    type ModelTy0.modelTy = ModelTy1.modelTy
  clone CreusotContracts_Logic_Model_Impl0_Model_Stub as Model0 with type t = Alloc_Vec_Vec_Type.t_vec t a1,
    type ModelTy0.modelTy = ModelTy0.modelTy
  val eq [@cfg:stackify] (self : Alloc_Vec_Vec_Type.t_vec t a1) (other : Alloc_Vec_Vec_Type.t_vec u a2) : bool
    ensures { result = (Model0.model self = Model1.model other) }
    
end
```

I haven't bisected yet, but this error is new from a few weeks ago (the code validates correctly on ea4a283193a00ec0583d37e13f04e279a68bb736)

**Comment by xldenis** [comment]
The only PR which makes sense for this would be https://github.com/xldenis/creusot/pull/586

**Comment by gmorenz** [comment]
Yep, that commit is where the issue first starts appearing, thanks for the pointer :)

**Comment by xldenis** [comment]
Yea this is a particularly tricky part of Creusot, one on which I've stumbled several times. Seems like we lose the predicate which lets us unify the types of both models. I would expect the issue to be in the logic of `param_env`.

**Comment by gmorenz** [comment]
I might still be misunderstanding things, but if not:

I think [this filter](https://github.com/xldenis/creusot/pull/586/files#diff-39195cd94c3cb6d0cd264773cbaf4bc2aefa40682720aef0571200e3139362f6R383) is wrong.  `predicate_may_hold_fatal` "Evaluates whether the obligation obligation can be satisfied (by any means).". I.e. it's evaluating `exists U: universe. obligation(U)` and the diff in question is filtering out predicates where that holds... which is to say that it's filtering out predicates that only *might* succeed, instead of filtering out predicates that *definitely do* succeed. The comment suggests it was intended to only do the latter, and filtering out predicates that might not succeed seems probably wrong to me.

If this filter is important, I'm not sure how to fix it.

My first thought was to try and negate the predicate and filter out predicates whose negations definitely fail. Since `!exists U. !obligation(U)` is the same as `forall U. obligation(U)` (and just allow all predicates we can't negate - since flip_polarity returns an `Option<Predicate>`). Unfortunately that doesn't work because

1. It causes a bunch of "could not resolve trait instance" panics - probably because we call methods on traits that are definitely implemented? If that's the case filtering at all seems suspect?

2. It filters out a bunch of stuff that *wasn't filtered previously*, i.e. `predicate_may_hold_fatal` returns false for both the original predicate, and it's negation. I suppose this might be valid if the set of types we're checking the obligation on is empty? Or if `predicate_may_hold_fatal` allows itself to return false negatives? Or maybe I misunderstand something, like `flip_polarity` doesn't actually return a logical negation predicate?

Coming back to "if this filter is important", I don't understand why this filter exists in the first place? It seems like if we just revert this piece of the change (add back `additional_predicates.extend(es.predicates_for(self.tcx, subst))` and remove the block that replaced it) everything just works?

**Comment by xldenis** [comment]
> Coming back to "if this filter is important", I don't understand why this filter exists in the first place? It seems like if we just revert this piece of the change (add back additional_predicates.extend(es.predicates_for(self.tcx, subst)) and remove the block that replaced it) everything just works?

If we do this we get a ton of other issues that #586 was solving because we introduce ambiguities. The `ParamEnv` stores the context in which trait resolution will be performed. If you have something like `[Vec<T> : Display, T : Display]` as your param env, and you ask to resolve `Display Vec<T>`, trait resolution will not actually select the *real* instance of `Display` for `Vec` but will provide the one coming from the `ParamEnv`. In Rust parlance, the instance will be `ImplSource::Param` and not `ImplSource::UserDefined`. This is a major problem. The idea of the fix in #586 was to filter out those predicates which could introduce ambiguity in trait resolution.

You're right that `may_hold_fatal` is wrong in hindsight because of that quantification issue. I'm not sure how to best fix this though. Perhaps, we can attempt to add a predicate, perform resolution and ask if there is an ambiguity? Or perhaps what we need to do is just run `resolve_opt` instead of building a seleciton context by hand. 



ref, rust-lang zulip chat about this issue: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/.E2.9C.94.20Normalize.20a.20ParamEnv

**Comment by xldenis** [comment]
Hmm. I think that running `resolve_opt` and checking if the result is `None` is probably the solution, it runs the same logic as Rust for trait resolution

**Comment by xldenis** [comment]
Ah ha! I think I've actually solved the issue. It is actually correct that Why3 refuses to type this, the real issue is that Creusot doesn't error earlier.

The problem is that the model instance for `Vec<T>` has `ModelTy = Seq<T>`. This means that `@self : Seq<T>` but the Eq instance has `Vec<U>` on the right-hand side, which has `@rhs : Seq<U>`. This is similar to the problem of the `PartialEq` instance for `Option`, and is actually solved in #613.

**Comment by xldenis** [comment]
I think the real question here is when / how Creusot should emit a type checking error about the contract not typing properly.

**Comment by xldenis** [comment]
Confirmed to be fixed by #613.
